src/Workspaces/CSharp/Portable/Formatting/CSharpSyntaxFormattingService.cs

```diff
@@ -1,63 +1,64 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.Composition;
-using System.Linq;
 using System.Threading;
-using System.Threading.Tasks;
-using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Formatting;
 using Microsoft.CodeAnalysis.Formatting.Rules;
-using Microsoft.CodeAnalysis.Host.Mef;
-using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Shared.Collections;
-using Microsoft.CodeAnalysis.Shared.Utilities;
 using Microsoft.CodeAnalysis.Text;
-using Roslyn.Utilities;
+
+#if !CODE_STYLE
+using System;
+using System.Composition;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+#endif
 
 namespace Microsoft.CodeAnalysis.CSharp.Formatting
 {
+#if !CODE_STYLE
     [ExportLanguageService(typeof(ISyntaxFormattingService), LanguageNames.CSharp), Shared]
+#endif
     internal class CSharpSyntaxFormattingService : AbstractSyntaxFormattingService
     {
-        private readonly Lazy<IEnumerable<IFormattingRule>> _lazyExportedRules;
+        private readonly ImmutableList<IFormattingRule> _rules;
 
+#if !CODE_STYLE
         [ImportingConstructor]
-        public CSharpSyntaxFormattingService([ImportMany] IEnumerable<Lazy<IFormattingRule, OrderableLanguageMetadata>> rules)
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+#endif
+        public CSharpSyntaxFormattingService()
         {
-            _lazyExportedRules = new Lazy<IEnumerable<IFormattingRule>>(() =>
-                ExtensionOrderer.Order(rules)
-                                .Where(x => x.Metadata.Language == LanguageNames.CSharp)
-                                .Select(x => x.Value)
-                                .Concat(new DefaultOperationProvider())
-                                .ToImmutableArray());
+            _rules = ImmutableList.Create<IFormattingRule>(
+                new WrappingFormattingRule(),
+                new SpacingFormattingRule(),
+                new NewLineUserSettingFormattingRule(),
+                new IndentUserSettingsFormattingRule(),
+                new ElasticTriviaFormattingRule(),
+                new EndOfFileTokenFormattingRule(),
+                new StructuredTriviaFormattingRule(),
+                new IndentBlockFormattingRule(),
+                new SuppressFormattingRule(),
+                new AnchorIndentationFormattingRule(),
+                new QueryExpressionFormattingRule(),
+                new TokenBasedFormattingRule(),
+                new DefaultOperationProvider());
```

Comment:
I don't like the hardcoding here, unless we can add a unit test that is guaranteed to fail when we add a new `IFormattingRule` and don't update this list.

Reply:
> I don't like the hardcoding here, unless we can add a unit test that is guaranteed to fail when we add a new IFormattingRule and don't update this list.

There is no duplication of this list. A formatting rule does not exist unless it is in this list, so the test coverage associated with any new formatting rule guarantees it will not be omitted.

>  Are we trying to avoid MEF in code style layer?

There is no MEF in the code style layer.

Reply:
> A formatting rule does not exist unless it is in this list

Cool, as long as there is a single place where we list these rules, it looks fine to me.

---
In reply to: [236802648](https://github.com/dotnet/roslyn/pull/31276#discussion_r236802648) [](ancestors = 236802648)